src/IR/DConToFunc.hs

```diff
@@ -0,0 +1,166 @@
+  insertArities :: (I.TConId, TypeDef Poly.Type) -> ArityEnv
+  insertArities (_, TypeDef { variants = vars }) = foldl
+    (\acc var -> acc <> insertArity var)
+    M.empty
+    vars
+   where
+    -- | Add single DCon to the arity environment
+    insertArity :: (I.DConId, TypeVariant Poly.Type) -> ArityEnv
+    insertArity (dconid, variant) =
+      M.insert dconid (variantFields variant) M.empty
+
+{- | 'dConToFunc' modifies programDefs and traverses the IR to accomplish two tasks:
+
+  1) Add top-level constructor functions for each non-nullary 'DCon' to progamDefs
+  2) Turn non-nullary data constuctors into calls to top level constructor funcs
+-}
+dConToFunc :: I.Program Poly.Type -> Compiler.Pass (I.Program Poly.Type)
+dConToFunc p@I.Program { I.programDefs = defs, I.typeDefs = tDefs } =
+  runArityFn $ do
+    defs' <- local (createArityEnv tDefs <>)
+                   (mapM (\(v, e) -> (v, ) <$> dataToApp e) defs)
+    return p { I.programDefs = defs' ++ concat (createFuncs <$> tDefs) }
+
+{- | Replace data constructor application with function application
+
+  Turn I.App instances of the form (I.Data _ _) into (I.Var _ _)
+-}
+dataToApp :: I.Expr Poly.Type -> ArityFn (I.Expr Poly.Type)
+dataToApp a@(I.Data dconid t) = do
+  Just arity <- asks (M.lookup dconid)
+  case arity of
+    0 -> return a -- leave nullary data constructors alone
+    _ -> return $ I.Var (I.VarId $ fromString $ "__" ++ ident dconid) t
+dataToApp (I.App left right t) =
+  I.App <$> dataToApp left <*> dataToApp right <*> pure t
+dataToApp (I.Let args body t) =
+  I.Let <$> mapM recurse args <*> dataToApp body <*> pure t
+  where recurse = \(b, e) -> (,) b <$> dataToApp e
+dataToApp (I.Lambda b e t) = I.Lambda b <$> dataToApp e <*> pure t
+dataToApp (I.Match scrut arms t) =
+  I.Match <$> dataToApp scrut <*> mapM recurse arms <*> pure t
+  where recurse = \(alt, e) -> (,) alt <$> dataToApp e
+dataToApp (I.Prim p es t) = I.Prim p <$> mapM dataToApp es <*> pure t
+dataToApp a               = pure a
+
+{- | Worked example of ADT definition and corresponding constructor functions
+
+type Shape 
+  Square Int 
+  Rect Int Int
+
+  Every top-level function has the form (I.VarId, I.Expr Poly.Type) = (functionName, functionBody)
+  The function body is a lambda expression representing a "call" to a fully applied data constructor.
+
+  Let's first consider the top-level func for Square:
+  (Square, body) 
+  body = fun arg0 { App L R t } : Int -> Shape
+   where L = Square : Int -> Shape
+         R = arg0 : type Int
+         t = Shape, because the type of a fully applied data constructor is its type constructor
+
+  Next consider the top-level func for Rect:
+  (Rect, body*) *Notice the body of the lambda contains another lambda (curried arguments),
+                 and the Application expression contains another Application (curried arguments).  
+  body = fun arg0 { fun arg1 { App L R t } : Int -> Shape } : Int -> Int -> Shape
```

Comment:
I added a description at the top without IR syntax, but I also kept the old description and put it under a section 'Representing constructor functions in the IR'. I think it might be helpful to keep because it goes over the types of nested lambdas and their arguments. If you think the IR syntax description is too confusing or not helpful, I will remove it. Let me know what you think.